test/e2e/network/traffic_distribution.go

```diff
@@ -119,164 +138,323 @@ var _ = common.SIGDescribe(feature.TrafficDistribution, framework.WithFeatureGat
+				},
+			)
+			break
+		}
 
-			requestsSucceedByReachingAnyServingPod := framework.MakeMatcher(func(reverseChronologicalLogLines []string) (func() string, error) {
-				logLines := reverseChronologicalLogLines
-				if len(logLines) < 20 {
-					return gomegaCustomError("got %d log lines, waiting for at least 20\nreverseChronologicalLogLines=\n%v", len(logLines), strings.Join(reverseChronologicalLogLines, "\n")), nil
-				}
+		doTrafficDistributionTest(ctx, trafficDist, clients, endpoints)
+	}
 
-				// Requests are counted as successful when the response read from the log
-				// lines is the name of a recognizable serving pod.
-				consecutiveSuccessfulRequests := 0
+	framework.It("should route traffic to an endpoint in the same zone when using PreferClose", func(ctx context.Context) {
+		doSimpleSameZoneTrafficDistributionTest(ctx, v1.ServiceTrafficDistributionPreferClose)
+	})
 
-				for _, logLine := range logLines {
-					if logLine == "" || strings.HasPrefix(logLine, "Date:") {
-						continue
-					}
-					_, servingPodExists := zoneForServingPod[logLine]
-					if !servingPodExists {
-						return gomegaCustomError("request from client pod likely failed because we got an unrecognizable response = %v; want response to be one of the serving pod names\nreverseChronologicalLogLines=\n%v", logLine, strings.Join(reverseChronologicalLogLines, "\n")), nil
-					}
-					consecutiveSuccessfulRequests++
-					if consecutiveSuccessfulRequests >= 10 {
-						return nil, nil // Pass condition
-					}
-				}
-				// Ideally, the matcher would never reach this condition
-				return gomegaCustomError("requests didn't meet the required criteria to reach a serving pod\nreverseChronologicalLogLines=\n%v", strings.Join(reverseChronologicalLogLines, "\n")), nil
-			})
+	framework.It("should route traffic correctly between pods on multiple nodes when using PreferClose", func(ctx context.Context) {
+		doMultiNodeSameZoneTrafficDistributionTest(ctx, v1.ServiceTrafficDistributionPreferClose)
+	})
 
-			gomega.Eventually(ctx, requestsFromClient(clientPod)).WithPolling(5 * time.Second).WithTimeout(e2eservice.KubeProxyLagTimeout).Should(requestsSucceedByReachingAnyServingPod)
+	framework.It("should route traffic to an endpoint in the same zone when using PreferSameZone", framework.WithFeatureGate(features.PreferSameTrafficDistribution), func(ctx context.Context) {
+		doSimpleSameZoneTrafficDistributionTest(ctx, v1.ServiceTrafficDistributionPreferSameZone)
+	})
 
-		})
+	framework.It("should route traffic currently between pods on multiple nodes when using PreferSameZone", framework.WithFeatureGate(features.PreferSameTrafficDistribution), func(ctx context.Context) {
+		doMultiNodeSameZoneTrafficDistributionTest(ctx, v1.ServiceTrafficDistributionPreferSameZone)
+	})
 
+	framework.It("should route traffic to an endpoint on the same node when using PreferSameNode", framework.WithFeatureGate(features.PreferSameTrafficDistribution), func(ctx context.Context) {
```

Comment:
I think that the most important part of this feature is
- prefer same node if available
- failover when not available
- when same node is available again prefer same node

This test must guarantee that implementations meet this behavior and is a serious candidate to conformance.
I would not worry about duplicating code if you feel is a stretch to reuse the helper, but I think is important we ensure in this test these 3 things are correct

Reply:
nevermind, I got it now, you are testing that with a third client,

Reply:
I'm going to delete this comment, I should have read the test entirely first, sorry